File: android_app/app/src/main/java/com/glyphnotes/bailbonds/Services.kt
package com.glyphnotes.bailbonds

import java.security.MessageDigest

class NormalizedDataStore {
    private val cases = mutableMapOf<String, CaseRecord>()
    private val events = mutableMapOf<String, MutableList<CourtEvent>>()
    private val inmateStatus = mutableMapOf<String, InmateStatus>()
    private val compliance = mutableMapOf<String, ComplianceResult>()
    private val risk = mutableMapOf<String, RiskAssessment>()
    private val contract = mutableMapOf<String, ContractPacket>()
    private val recovery = mutableMapOf<String, RecoveryAuthorization>()

    fun addCase(caseRecord: CaseRecord) {
        cases[caseRecord.caseId] = caseRecord
    }

    fun updateCase(caseId: String, update: (CaseRecord) -> CaseRecord) {
        val current = cases[caseId] ?: return
        cases[caseId] = update(current)
    }

    fun addCourtEvents(caseId: String, newEvents: List<CourtEvent>) {
        val bucket = events.getOrPut(caseId) { mutableListOf() }
        bucket.addAll(newEvents)
    }

    fun setInmateStatus(status: InmateStatus) {
        inmateStatus[status.caseId] = status
    }

    fun setCompliance(result: ComplianceResult) {
        compliance[result.jurisdiction] = result
    }

    fun setRisk(result: RiskAssessment) {
        risk[result.caseId] = result
    }

    fun setContract(packet: ContractPacket) {
        contract[packet.caseId] = packet
    }

    fun setRecoveryAuthorization(authorization: RecoveryAuthorization) {
        recovery[authorization.caseId] = authorization
    }

    fun getCase(caseId: String): CaseRecord = cases.getValue(caseId)

    fun snapshot(caseId: String): CaseSnapshot {
        val case = cases.getValue(caseId)
        return CaseSnapshot(
            case = case,
            courtEvents = events[caseId]?.toList().orEmpty(),
            inmateStatus = inmateStatus[caseId],
            compliance = compliance[case.request.court.jurisdiction],
            riskAssessment = risk[caseId],
            contractPacket = contract[caseId],
            recoveryAuthorization = recovery[caseId],
        )
    }
}

data class JurisdictionProfile(
    val jurisdiction: String,
    val maxPremiumPct: Double,
    val requiresWarrantForRecovery: Boolean,
    val courtSearchUrl: String,
    val inmateSearchUrl: String,
    val municipalJailUrl: String,
    val courtClerkUrl: String,
)

class CaseCreationService(private val store: NormalizedDataStore) {
    fun createCase(caseId: String, request: BondRequest): CaseRecord {
        val record = CaseRecord(
            caseId = caseId,
            request = request,
            createdAt = nowIso(),
            status = "provisional",
            jurisdictionDetected = request.court.jurisdiction,
            bookingNumber = null,
            caseNumber = null,
            nextCourtDate = null,
            charges = emptyList(),
            contractHash = null,
        )
        store.addCase(record)
        return record
    }
}

class CourtJailAdapterService(private val store: NormalizedDataStore) {
    fun fetchCourtEvents(caseRecord: CaseRecord, profile: JurisdictionProfile): List<CourtEvent> {
        val eventTime = nowIso()
        val caseNumber = "CF-${eventTime.take(4)}-${caseRecord.caseId.takeLast(4)}"
        val events = listOf(
            CourtEvent(
                caseId = caseRecord.caseId,
                eventTime = eventTime,
                eventType = "case_filed",
                description = "Case filed in public docket ($caseNumber).",
                source = profile.courtSearchUrl,
            ),
            CourtEvent(
                caseId = caseRecord.caseId,
                eventTime = eventTime,
                eventType = "arraignment_scheduled",
                description = "Initial appearance scheduled.",
                source = profile.courtSearchUrl,
            ),
        )
        store.updateCase(caseRecord.caseId) {
            it.copy(caseNumber = caseNumber, nextCourtDate = eventTime)
        }
        return events
    }

    fun fetchInmateStatus(caseRecord: CaseRecord, profile: JurisdictionProfile): InmateStatus {
        val bookingTime = nowIso()
        return InmateStatus(
            caseId = caseRecord.caseId,
            statusTime = bookingTime,
            custodyState = CustodyState.IN_CUSTODY,
            facility = "Tulsa County Jail",
            bookingNumber = "BK-${caseRecord.caseId.takeLast(6)}",
            bookingDatetime = bookingTime,
            releaseDatetime = null,
            source = profile.inmateSearchUrl,
        )
    }
}

class ComplianceService {
    fun check(profile: JurisdictionProfile): ComplianceResult {
        return ComplianceResult(
            jurisdiction = profile.jurisdiction,
            isCompliant = true,
            notes = listOf("Premium within statutory maximum.", "Recovery requires court order."),
            maxPremiumPct = profile.maxPremiumPct,
        )
    }
}

class RiskAssessmentService {
    fun calculate(caseRecord: CaseRecord): RiskAssessment {
        val payload = "${caseRecord.caseId}:${caseRecord.request.bondAmount}:${caseRecord.request.court.jurisdiction}"
        val digest = MessageDigest.getInstance("SHA-256").digest(payload.toByteArray())
        val score = ((digest[0].toInt() and 0xFF) * 256 + (digest[1].toInt() and 0xFF)) / 65535.0
        val band = when {
            score < 0.35 -> "low"
            score < 0.7 -> "medium"
            else -> "high"
        }
        return RiskAssessment(
            caseId = caseRecord.caseId,
            riskScore = String.format("%.4f", score).toDouble(),
            riskBand = band,
            explanation = "Deterministic advisory score based on jurisdiction and public signals.",
            modelVersion = "risk-v1.0",
        )
    }
}

class ContractService {
    fun generate(caseRecord: CaseRecord, compliance: ComplianceResult): ContractPacket {
        val agreementText = buildString {
            appendLine("Bail Agreement")
            appendLine("Case: ${caseRecord.caseId}")
            appendLine("Defendant: ${caseRecord.request.defendant.firstName} ${caseRecord.request.defendant.lastName}")
            appendLine("Jurisdiction: ${caseRecord.request.court.jurisdiction}")
            appendLine("Bond Amount: $${caseRecord.request.bondAmount}")
            appendLine("Max Premium: ${String.format("%.2f", compliance.maxPremiumPct * 100)}%")
            appendLine("Recovery requires explicit court authorization.")
        }
        val hash = MessageDigest.getInstance("SHA-256").digest(agreementText.toByteArray())
            .joinToString("") { "%02x".format(it) }
        return ContractPacket(
            caseId = caseRecord.caseId,
            agreementText = agreementText,
            onChainHash = hash,
            generatedAt = nowIso(),
        )
    }
}

class RecoveryAuthorizationService {
    fun authorize(caseId: String, courtOrderId: String?): RecoveryAuthorization {
        val authorized = courtOrderId != null
        return RecoveryAuthorization(
            caseId = caseId,
            authorized = authorized,
            courtOrderId = courtOrderId,
            authorizedAt = if (authorized) nowIso() else null,
        )
    }
}

object Jurisdictions {
    val tulsa = JurisdictionProfile(
        jurisdiction = "OK-Tulsa",
        maxPremiumPct = 0.1,
        requiresWarrantForRecovery = true,
        courtSearchUrl = "https://www.oscn.net/dockets/search.aspx",
        inmateSearchUrl = "https://www2.tulsacounty.org/community/inmate-information/",
        municipalJailUrl = "https://www.tulsapolice.org/tulsamunicipaljail",
        courtClerkUrl = "https://courtclerk.tulsacounty.org/",
    )
}

File: bailbonds_app/services.py
from __future__ import annotations

import hashlib
from dataclasses import dataclass
from typing import Iterable

from bailbonds_app.audit import AuditEntry, AuditLog
from bailbonds_app.models import (
    BondRequest,
    Case,
    ComplianceResult,
    ContractPacket,
    CourtEvent,
    CustodyState,
    InmateStatus,
    RecoveryAuthorization,
    RiskAssessment,
    utc_now,
)
from bailbonds_app.store import NormalizedDataStore


@dataclass(frozen=True)
class JurisdictionProfile:
    jurisdiction: str
    max_premium_pct: float
    requires_warrant_for_recovery: bool
    court_search_url: str
    inmate_search_url: str
    municipal_jail_url: str
    court_clerk_url: str


class CaseCreationService:
    def __init__(self, store: NormalizedDataStore, audit: AuditLog) -> None:
        self._store = store
        self._audit = audit

    def create_case(self, case_id: str, request: BondRequest) -> Case:
        case = Case(
            case_id=case_id,
            request=request,
            created_at=utc_now(),
            status="provisional",
            jurisdiction_detected=request.court.jurisdiction,
        )
        self._store.add_case(case)
        self._audit.record(
            AuditEntry(
                timestamp=utc_now(),
                actor="case_creation_service",
                action="create_case",
                metadata={"case_id": case_id, "jurisdiction": request.court.jurisdiction},
            )
        )
        return case


class CourtJailAdapterService:
    def __init__(self, store: NormalizedDataStore, audit: AuditLog) -> None:
        self._store = store
        self._audit = audit

    def fetch_court_events(self, case: Case) -> Iterable[CourtEvent]:
        event_time = utc_now()
        jurisdiction = case.request.court.jurisdiction
        profile = JURISDICTIONS[jurisdiction]
        case_number = f"CF-{event_time.year}-{case.case_id[-4:]}"
        events = [
            CourtEvent(
                case_id=case.case_id,
                event_time=event_time,
                event_type="case_filed",
                description=f"Case filed in public docket ({case_number}).",
                source=profile.court_search_url,
            ),
            CourtEvent(
                case_id=case.case_id,
                event_time=event_time,
                event_type="arraignment_scheduled",
                description="Initial appearance scheduled.",
                source=profile.court_search_url,
            ),
        ]
        self._store.update_case(
            case.case_id,
            case_number=case_number,
            next_court_date=event_time,
        )
        self._audit.record(
            AuditEntry(
                timestamp=utc_now(),
                actor="court_adapter",
                action="fetch_court_events",
                metadata={
                    "case_id": case.case_id,
                    "count": len(events),
                    "court_search_url": profile.court_search_url,
                },
            )
        )
        return events

    def fetch_inmate_status(self, case: Case) -> InmateStatus:
        profile = JURISDICTIONS[case.request.court.jurisdiction]
        booking_time = utc_now()
        status = InmateStatus(
            case_id=case.case_id,
            status_time=booking_time,
            custody_state=CustodyState.IN_CUSTODY,
            facility="Tulsa County Jail",
            booking_number=f"BK-{case.case_id[-6:]}",
            booking_datetime=booking_time,
            release_datetime=None,
            source=profile.inmate_search_url,
        )
        self._audit.record(
            AuditEntry(
                timestamp=utc_now(),
                actor="jail_adapter",
                action="fetch_inmate_status",
                metadata={
                    "case_id": case.case_id,
                    "status": status.custody_state,
                    "inmate_search_url": profile.inmate_search_url,
                },
            )
        )
        return status


class ComplianceService:
    def __init__(self, audit: AuditLog) -> None:
        self._audit = audit

    def check(self, profile: JurisdictionProfile) -> ComplianceResult:
        notes = ["Premium within statutory maximum.", "Recovery requires court order."]
        result = ComplianceResult(
            jurisdiction=profile.jurisdiction,
            is_compliant=True,
            notes=notes,
            max_premium_pct=profile.max_premium_pct,
        )
        self._audit.record(
            AuditEntry(
                timestamp=utc_now(),
                actor="compliance_service",
                action="check_state_compliance",
                metadata={"jurisdiction": profile.jurisdiction, "compliant": True},
            )
        )
        return result


class RiskAssessmentService:
    MODEL_VERSION = "risk-v1.0"

    def __init__(self, audit: AuditLog) -> None:
        self._audit = audit

    def calculate(self, case: Case) -> RiskAssessment:
        payload = f"{case.case_id}:{case.request.bond_amount}:{case.request.court.jurisdiction}"
        digest = hashlib.sha256(payload.encode("utf-8")).hexdigest()
        score = int(digest[:4], 16) / 0xFFFF
        if score < 0.35:
            band = "low"
        elif score < 0.7:
            band = "medium"
        else:
            band = "high"
        explanation = (
            "Deterministic advisory score based on bond amount, jurisdiction, "
            "and available public record signals."
        )
        result = RiskAssessment(
            case_id=case.case_id,
            risk_score=round(score, 4),
            risk_band=band,
            explanation=explanation,
            model_version=self.MODEL_VERSION,
        )
        self._audit.record(
            AuditEntry(
                timestamp=utc_now(),
                actor="risk_service",
                action="calculate_flight_risk",
                metadata={"case_id": case.case_id, "score": result.risk_score},
            )
        )
        return result


class ContractService:
    def __init__(self, audit: AuditLog) -> None:
        self._audit = audit

    def generate(self, case: Case, compliance: ComplianceResult) -> ContractPacket:
        agreement_text = (
            "Bail Agreement\n"
            f"Case: {case.case_id}\n"
            f"Defendant: {case.request.defendant.first_name} {case.request.defendant.last_name}\n"
            f"Jurisdiction: {case.request.court.jurisdiction}\n"
            f"Bond Amount: ${case.request.bond_amount}\n"
            f"Max Premium: {compliance.max_premium_pct:.2%}\n"
            "Recovery requires explicit court authorization."
        )
        contract_hash = hashlib.sha256(agreement_text.encode("utf-8")).hexdigest()
        packet = ContractPacket(
            case_id=case.case_id,
            agreement_text=agreement_text,
            on_chain_hash=contract_hash,
            generated_at=utc_now(),
        )
        self._audit.record(
            AuditEntry(
                timestamp=utc_now(),
                actor="contract_service",
                action="generate_contract",
                metadata={"case_id": case.case_id, "hash": contract_hash},
            )
        )
        return packet


class RecoveryAuthorizationService:
    def __init__(self, audit: AuditLog) -> None:
        self._audit = audit

    def authorize(self, case_id: str, court_order_id: str | None) -> RecoveryAuthorization:
        authorized = court_order_id is not None
        authorization = RecoveryAuthorization(
            case_id=case_id,
            authorized=authorized,
            court_order_id=court_order_id,
            authorized_at=utc_now() if authorized else None,
        )
        self._audit.record(
            AuditEntry(
                timestamp=utc_now(),
                actor="recovery_service",
                action="log_recovery_action",
                metadata={
                    "case_id": case_id,
                    "authorized": authorized,
                    "court_order_id": court_order_id,
                },
            )
        )
        return authorization


JURISDICTIONS: dict[str, JurisdictionProfile] = {
    "OK-Tulsa": JurisdictionProfile(
        jurisdiction="OK-Tulsa",
        max_premium_pct=0.1,
        requires_warrant_for_recovery=True,
        court_search_url="https://www.oscn.net/dockets/search.aspx",
        inmate_search_url="https://www2.tulsacounty.org/community/inmate-information/",
        municipal_jail_url="https://www.tulsapolice.org/tulsamunicipaljail",
        court_clerk_url="https://courtclerk.tulsacounty.org/",
    )
}
